moggieuk/Happy-Hare
# Allow tracking of extrusion multiplier per tool

**Issue:**
When running a multi-color print I had to adjust the extrusion multiplier mid-print for a single filament.  The remainder of the print needed to either be watched to reset the value for each tool change or allow all tools to use the modified value.

**Fix Description:**
This request adds the ability for HappyHare to track changes to the extrusion multipliers per tool and apply them on tool change.  This optional feature is enabled by setting the `enable_tool_extrusion_multiplier` value in `mmu_parameters.cfg`.   The default setting is disabled.  Once set after a tool is unloaded the current extrusion multiplier is retrieved from the `gcode_move` module and stored in a list.   On tool load the last value is retrieved from the list and applied by sending an `M221` command.   This list is reset to 100% for all tools on either a `FIRMWARE_RESTART` or Klipper restart.

A new entry has also been added to the API `get_status` method
Configuration example has been added to `mmu_parameters.cfg` in the optional section.

**Testing Done**
This has been tested with a live print as well as the manual steps listed below.
1. Home printer and MMU
2. Load a tool and set the extrusion multiplier
3. Load a second tool and verify
  3a. The extrusion multiplier reset to 100%
  3b. `get_status` returned the value set in step 2 for the initial tool
4. Load the tool loaded in step 2 and verify:
  4a. The extrusion multiplier was set to the value set in step 2
  4b. `get_status` returned the value set in step 2

**GitHub Issues Resolved**
None
